src/workerd/api/streams/internal.c++

```diff
@@ -175,18 +175,18 @@ public:
     // pump operation's backpressure. So we can't use the default `ReadableStreamSource::pumpTo()`
     // implementation, and have to implement our own.
 
-    auto outputAdapter = kj::heap<PumpAdapter>(output);
-    auto promise = inner->pumpTo(*outputAdapter).attach(kj::mv(outputAdapter)).ignoreResult();
+    PumpAdapter outputAdapter(output);
+    co_await inner->pumpTo(outputAdapter);
 
     if (end) {
-      promise = promise.then([&output]() { return output.end(); });
+      co_await output.end();
     }
 
     // We only use `TeeBranch` when a locally-sourced stream was tee'd (because system streams
     // implement `tryTee()` in a different way that doesn't use `TeeBranch`). So, we know that
     // none of the pump can be performed without the IoContext active, and thus
     // `DeferredProxy` has to be a noop.
-    return addNoopDeferredProxy(kj::mv(promise));
+    co_return newNoopDeferredProxy();
```

Comment:
This changes the timing for the deferred proxy (I hit this trap myself in blob.c++). The actual pumpTo and end need to be contained in an inner promise that is held by the deferred proxy. So, you'll want something like...

Reply:
There are two promises here... the outer one for the `DeferredProxy` itself, and the inner one for the pump. With this, the outer promise for the `DeferredProxy` itself gets delayed until after the inner promise is resolved, whereas before it would be resolved immediately with the inner promise still pending, which is what we want. As I understand it, the noop here is that the outer promise isn't actually waiting for anything, not that the inner promise is already resolved.

Reply:
> whereas before it would be resolved immediately with the inner promise still pending

I don't think that's true. `addNoopDeferredProxy` awaits the given promise first, _then_ returns a `DeferredProxy` that is already resolved (that's what makes it a "noop" DeferredProxy). The code Mike has written here should have the same behavior.

Same thing goes for the changes to `blob.c++` in #921, I don't see the difference between the old code and the new.

Neither of these scenarios are actually eligible for deferred proxying AFAICT.

Reply:
Did you actually see a test failure caused by your original change to `blob.c++`, which is resolved by #921?

Reply:
hmm, I think you're right. I'd forgotten that `addNoopDeferredProxy` waited that promise. We should add some more documentation around this because it's a bit of a mind bender. Or, perhaps even better, update `DeferredProxy` itself to make this clearer. Specifically, if the `DeferredProxy` is a no-op, then perhaps the inner promise should be nullptr rather than an already resolved promise (e.g. make it `kj::Maybe<kj::Promise<void>>` instead. This would make it clearer that there's nothing else to wait on once the `kj::Promise<DeferredProxy>` is fulfilled.